include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,1012 @@
+        : p_(str.data())
+        , n_(str.size())
+        , dn_(pct_decode_bytes_unchecked(str))
+        , plus_to_space_(plus_to_space)
+    {}
+
+#if !defined(BOOST_NO_CXX17_HDR_STRING_VIEW)
+    /** Constructor
+     */
+    decoded_view(
+        std::string_view str,
+        bool plus_to_space = true) noexcept
+        : p_( str.data() )
+        , n_( str.size() )
+        , dn_(pct_decode_bytes_unchecked(str))
+        , plus_to_space_(plus_to_space)
+    {}
+#endif
+
+    /** Returns an iterator to the beginning
+
+        Returns an iterator to the first decoded character of the string
+
+        @return const_iterator to the first character
+    */
+    BOOST_URL_DECL
+    const_iterator
+    begin() const noexcept;
+
+    /// @copydoc begin()
+    BOOST_URL_DECL
+    const_iterator
+    cbegin() const noexcept;
+
+    /** Returns an iterator to the end
+
+        Returns an iterator to one past the last decoded character of the string.
+
+        This character acts as a placeholder, attempting to access it
+        results in undefined behavior.
+
+        @return const_iterator to the character following the last character
+    */
+    BOOST_URL_DECL
+    const_iterator
+    end() const noexcept;
+
+    /// @copydoc end()
+    BOOST_URL_DECL
+    const_iterator
+    cend() const noexcept;
+
+    /** Accesses the first character
+
+        Returns the decoded first character in the string.
+
+        The behavior is undefined if `empty() == true`.
+
+        @return reference to the first character
+     */
+    BOOST_URL_DECL
+    const_reference
+    front() const;
+
+    /** Accesses the last character
+
+        Returns the last decoded character in the string
+
+        The behavior is undefined if `empty() == true`.
+
+        @return reference to the first character
+     */
+    BOOST_URL_DECL
+    const_reference
+    back() const;
+
+    /** Returns a pointer to the first encoded character of a string
+
+        Returns a pointer to the underlying character array.
+
+        The pointer is such that the range `[data(); data() + encoded_size())`
+        is valid and the values in it correspond to the values of the
+        string.
+
+        @note Unlike `std::basic_string::data()` and string literals,
+        `data()` may return a pointer to a buffer that is not
+        null-terminated. Therefore, it is typically a mistake to pass
+        data() to a routine that takes just a const char* and expects
+        a null-terminated string.
+
+        @return A pointer to the underlying character array.
+     */
+    BOOST_CONSTEXPR
+    const char*
+    encoded_data() const
+    {
+        return p_;
+    }
+
+    /** Returns the number of char elements in the string
+
+        @return The number of char elements in the string
+    */
+    BOOST_CONSTEXPR
+    size_type
+    size() const noexcept
+    {
+        return dn_;
+    }
+
+    /// @copydoc size()
+    BOOST_CONSTEXPR
+    size_type
+    length() const noexcept
```

Comment:
Why two versions of the same function?

Reply:
I'm modeling the class after `string_view`. At first, it's kind of the point that replacing `string_view` with a `decoded_view` should be as easy as possible.

Reply:
No one uses `string_view::length` We should not repeat the mistakes made by other authors. Prefer typing less rather than more. "You ain't gonna need it." Let users complain first, that `decoded_view` is hard to drop-in as a replacement for `string_view`, before adding the kitchen sink. Use a light touch, just add what is urgent and leave the rest for another day. No one is using reverse iterators bro! Except maybe for college students and people trying to get rich writing C++ books.

Look:

![image](https://user-images.githubusercontent.com/1503976/178654883-993c0bc0-964d-46e1-ba79-544d5181a59b.png)

![image](https://user-images.githubusercontent.com/1503976/178654913-1404e394-794f-4265-a3b8-13f2555f7bf2.png)